File: typescript-fp/src/chess/ChessBoard.ts
import {Color, other} from './Color'
import {filter, flatten, getOrElse, map, none, Option, some} from 'fp-ts/Option'
import {charRangeInclusive, range, updatedList} from "./utils";
import {chunksOf, replicate, zip} from "fp-ts/ReadonlyArray";
import {Position, toIndex, valid} from "./Position";
import {pipe} from 'fp-ts/lib/function'
import {takeLeft} from "fp-ts/Array";
import {Either, left as eitherLeft, right as eitherRight} from "fp-ts/Either";

export enum PieceType {
    Pawn = 1,
    Knight = 2,
    Bishop = 3,
    Rook = 4,
    Queen = 5,
    King = 6,
}

export interface Piece {
    readonly color: Color
    readonly type: PieceType
}

const showPiece = (piece: Piece): string => {
    let {color, type} = piece
    if(color == Color.White) {
        switch (type) {
            case PieceType.Pawn: return "♙"
            case PieceType.Knight: return "♘"
            case PieceType.Bishop: return "♗"
            case PieceType.Rook: return "♖"
            case PieceType.Queen: return  "♕"
            case PieceType.King: return "♔"
        }
    } else {
        switch (type) {
            case PieceType.Pawn: return "♟"
            case PieceType.Knight: return "♞"
            case PieceType.Bishop: return "♝"
            case PieceType.Rook: return "♜"
            case PieceType.Queen: return "♛"
            case PieceType.King: return "♚"
        }
    }
}

const ANSI_RESET = "\u001B[0m"

export enum BackgroundColor {
    Yellow = "\u001B[43m",
    Maroon = "\u001B[41m",
    Cyan = "\u001B[47m",
    White = "\u001B[46m",
}

export interface ChessBoard {
    toVector(): Option<Piece>[]
    highlights(): Option<BackgroundColor>[]
    nextPlayer(): Color
}

const piece = (color: Color, type: PieceType): Piece => ({color, type})

const noHighlight = (): Option<BackgroundColor>[] => range(0,63).map(_=> none)

export function initialPosition(): ChessBoard {
    const rearRow = [PieceType.Rook, PieceType.Knight, PieceType.Bishop, PieceType.Queen, PieceType.King, PieceType.Bishop,
        PieceType.Knight, PieceType.Rook]
    const whiteRearRow = rearRow.map(it => some(piece(Color.White, it)))
    const whiteFrontRow = range(0, 7).map(_ => some(piece(Color.White, PieceType.Pawn)))
    const emptyRow = range(0, 7).map(_ => none)

    const blackFrontRow = range(0, 7).map(_ => some(piece(Color.Black, PieceType.Pawn)))
    const blackRearRow = rearRow.map(it => some(piece(Color.Black, it)))

    const cells = whiteRearRow.concat(
        whiteFrontRow,
        range(0, 3).map(_ => emptyRow).flat(),
        blackFrontRow,
        blackRearRow
    )

    return {
        toVector(): Option<Piece>[] {
            return cells
        },
        highlights(): Option<BackgroundColor>[] {
            return noHighlight()
        },
        nextPlayer(): Color {
            return Color.White
        }
    }
}


export function showChessBoard(cb: ChessBoard): string {
    const columnHeader = charRangeInclusive('a', 'h').map(it => ` ${it} `).join("")
    const rowHeader = range(1, 8)
    const alternateBlackWhite = replicate(4,
                                range(0,3).map(_ => [BackgroundColor.White, BackgroundColor.Cyan])
                        .concat(range(0,3).map(_ => [BackgroundColor.Cyan,  BackgroundColor.White]))
    ).flat().flat()

    const colorCodes = zip(cb.highlights(), alternateBlackWhite)
    const cells = zip(colorCodes, cb.toVector())

    const cells2d  = chunksOf(8)(cells)

    function showLineCells(highlightCol_bgColor_piece: readonly [readonly [Option<BackgroundColor>, BackgroundColor], Option<Piece>]): string {
        const [color, piece] = highlightCol_bgColor_piece
        const [highlightCol, bgColor] = color
        return getOrElse(() => bgColor)(highlightCol) + " " + getOrElse(() => " ")(map(showPiece)(piece)) + " " + ANSI_RESET
    }

    const lines = cells2d.map((line) => line.map(showLineCells))

    const concatTuples = (xy: readonly [number, string[]]): string => xy[0] + " " + xy[1].join("") + " "+ xy[0]

    return `  ${columnHeader}\n${zip(rowHeader, lines).map(concatTuples).join("\n")}\n  ${columnHeader}`
}

function at(cb: ChessBoard, pos: Position): Option<Piece> {
    if(!valid(pos)) return none
    return cb.toVector()[toIndex(pos)]
}

export function cellIsEmpty(cb: ChessBoard, pos: Position): boolean {
    return at(cb, pos) == none
}

export function diffColor(cb: ChessBoard, x: Position, y: Position): boolean {
    function seeIfYIsValid(pieceX: Piece) {
        return pipe(
            at(cb, y),
            map(pieceY => pieceX.color != pieceY.color),
            getOrElse(()=> false)  //no piece at position y
        )
    }

    return getOrElse(()=> false)( //no piece at position x
        pipe(
            at(cb, x),
            map(seeIfYIsValid)
        ))
}


export function boardForTesting(pieces: Array<[Position, Piece]>): ChessBoard {
    const initial: Option<Piece>[] = range(0,63).map(_=> none)

    //UNSAFE: Mutable operation, used for testing only
    pieces.forEach(it => {
        initial[toIndex(it[0])] = some(it[1])
    })

    return {
        toVector(): Option<Piece>[] {
            return initial
        },
        highlights(): Option<BackgroundColor>[] {
            return noHighlight()
        },
        nextPlayer(): Color {
            return Color.White
        }
    }
}

function moveRecursion(cb: ChessBoard, start: Position, here: Position, fn: Function): Position[] {
    const next = fn(here)
    if(!valid(next)) return []
    if(cellIsEmpty(cb, next)) return [next].concat(moveRecursion(cb, start, next, fn))
    if(diffColor(cb, start, next)) return [next]
    return []
}

function movesFrom(cb: ChessBoard, here: Position, directionFn: Function) {
    return moveRecursion(cb, here, here, directionFn)
}

const up = (pos: Position): Position => { const [i, j] = pos; return [i-1, j]}
const down = (pos: Position): Position => { const [i, j] = pos; return [i+1, j]}
const left = (pos: Position): Position => { const [i, j] = pos; return [i, j-1]}
const right = (pos: Position): Position => { const [i, j] = pos; return [i, j+1]}
const diagUpLeft = (pos: Position): Position => { const [i, j] = pos; return [i-1, j-1]}
const diagUpRight = (pos: Position): Position => { const [i, j] = pos; return [i-1, j+1]}
const diagDownLeft = (pos: Position): Position => { const [i, j] = pos; return [i+1, j-1]}
const diagDownRight = (pos: Position): Position => { const [i, j] = pos; return [i+1, j+1]}

function rookMoves(cb: ChessBoard, pos: Position) {
    return movesFrom(cb, pos, up).concat(
        movesFrom(cb, pos, down),
        movesFrom(cb, pos, left),
        movesFrom(cb, pos, right)
    )
}

function kingMoves(cb: ChessBoard, pos: Position) {
    return takeLeft(1)(movesFrom(cb, pos, up)).concat(
        takeLeft(1)(movesFrom(cb, pos, down)),
        takeLeft(1)(movesFrom(cb, pos, left)),
        takeLeft(1)(movesFrom(cb, pos, right)),
        takeLeft(1)(movesFrom(cb, pos, diagUpLeft)),
        takeLeft(1)(movesFrom(cb, pos, diagUpRight)),
        takeLeft(1)(movesFrom(cb, pos, diagDownLeft)),
        takeLeft(1)(movesFrom(cb, pos, diagDownRight))
    )
}

function bishopMoves(cb: ChessBoard, pos: Position) {
    return movesFrom(cb, pos, diagUpLeft).concat(
        movesFrom(cb, pos, diagUpRight),
        movesFrom(cb, pos, diagDownLeft),
        movesFrom(cb, pos, diagDownRight),
    )
}

function queenMoves(cb: ChessBoard, pos: Position) {
    return rookMoves(cb, pos).concat(bishopMoves(cb, pos))
}

function knightMoves(cb: ChessBoard, pos: Position): Position[] {
    const fn = (y: Position) => diffColor(cb, pos, y) || cellIsEmpty(cb, y)

    return [pipe(pos, up, up, left)].concat(
        [pipe(pos, up, up, right)],
        [pipe(pos, down, down, left)],
        [pipe(pos, down, down, right)],
        [pipe(pos, left, left, up)],
        [pipe(pos, left, left, down)],
        [pipe(pos, right, right, up)],
        [pipe(pos, right, right, down)],
    ).filter(it => valid(it) && fn(it))
}

function pawnMovesNonCapture(cb: ChessBoard, pos: Position, dirFn: Function, movingForTheFirstTime: Function) {
    const isEmpty = (p: Position) => cellIsEmpty(cb, p)
    const next: Position = dirFn(pos)
    const nextToNext: Position = dirFn(next)

    if(movingForTheFirstTime(pos)){
        return [next].filter(isEmpty).concat(
            [nextToNext].filter(it => isEmpty(next) && isEmpty(it))
        )
    }

    return [next].filter(isEmpty)
}

function pawnMovesUp(cb: ChessBoard, pos: Position) {
    return pawnMovesNonCapture(cb, pos, up, (p: Position) => p[0] == 6).concat(
        [diagUpLeft(pos)].filter((y)=> diffColor(cb, pos, y)),
        [diagUpRight(pos)].filter((y: Position)=> diffColor(cb, pos, y))
    )
}

function pawnMovesDown(cb: ChessBoard, pos: Position) {
    return pawnMovesNonCapture(cb, pos, down, (p: Position) => p[0] == 1).concat(
        [diagDownLeft(pos)].filter((y)=> diffColor(cb, pos, y)),
        [diagDownRight(pos)].filter((y: Position)=> diffColor(cb, pos, y))
    )
}

//This is required for giving type hint; simply returning an empty array doesn't tell TS which type it is
function noMoves(): Position[] {
    return []
}

function possiblePositionsToMove(cb: ChessBoard, pos: Position): Position[] {
    function getPositions(piece: Piece): Position[] {
        const {color, type} = piece
        switch (type) {
            case PieceType.Bishop: return bishopMoves(cb, pos)
            case PieceType.Rook: return rookMoves(cb, pos)
            case PieceType.King: return kingMoves(cb, pos)
            case PieceType.Queen: return queenMoves(cb, pos)
            case PieceType.Knight: return knightMoves(cb, pos)
            case PieceType.Pawn: switch (color) {
                case Color.Black: return pawnMovesUp(cb, pos)
                case Color.White: return pawnMovesDown(cb, pos)
            }
        }
        return []
    }

    return getOrElse(noMoves)(pipe(
        at(cb, pos), map(getPositions)
    ))
}

export function showPossibleMoves(pos: Option<Position>, cb: ChessBoard): ChessBoard {
    function highlightedCells(p: Position): Option<BackgroundColor>[] {
        const cells = possiblePositionsToMove(cb, p).map(toIndex)
        return range(0, 63).map(x => cells.includes(x) ? some(BackgroundColor.Yellow) : none)
    }

    return pipe(pos,
               map(p => ({
                   toVector(): Option<Piece>[] {
                       return cb.toVector()
                   },
                   highlights(): Option<BackgroundColor>[] {
                       return highlightedCells(p)
                   },
                   nextPlayer(): Color {
                       return Color.White
                   }
               })),
               getOrElse(()=> cb))
}

export function movePiece(posX: Option<Position>, posY: Option<Position>, cb: ChessBoard): ChessBoard {
    function seeIfPosYIsValid(px: Position): ChessBoard {
        return pipe(posY,
                map(py => ({
                    toVector(): Option<Piece>[] {
                        const x = toIndex(px), y = toIndex(py);
                        return updatedList(cb.toVector(), [[x, none], [y, cb.toVector()[x]]])
                    },
                    highlights(): Option<BackgroundColor>[] {
                        return noHighlight()
                    },
                    nextPlayer(): Color {
                        return other(cb.nextPlayer())
                    }
                })),
                getOrElse(()=> cb)
            )
    }

    return pipe(posX,
            map(seeIfPosYIsValid),
            getOrElse(() => cb)
        )
}

export function validateAndMakeMove(posX: Option<Position>, posY: Option<Position>, cb: ChessBoard): Either<string[], ChessBoard> {
    const ifPlayersTurnMoveFrom = (x: Position) => (piece: Piece): Either<string[], ChessBoard> => {
        const {color, type} = piece
        if(color != cb.nextPlayer()) return eitherLeft(["Not your turn!"])

        const ys = possiblePositionsToMove(cb, x).map(it => toIndex(it))
        return pipe(posY,
                filter(py => ys.includes(toIndex(py))),
                map(py => {
                    return eitherRight(movePiece(posX, posY, cb))
                }),
                getOrElse(() => eitherLeft(["Target position not given or movement not allowed"]))
            )
    }

    function moveMaybePieceFromGivenPos(pos_piece: [Position, Option<Piece>]): Either<string[], ChessBoard> {
        return pipe(pos_piece[1],
            map(ifPlayersTurnMoveFrom(pos_piece[0])),
            getOrElse(()=> eitherLeft(["No piece at source position"]))
        )
    }

    return pipe(posX,
            map((px: Position): [Position, Option<Piece>] => [px, cb.toVector()[toIndex(px)]]),
            //Notice that type inference fails here, and we have to explicitly give return type from lambda
            map(moveMaybePieceFromGivenPos),
            getOrElse(() => eitherLeft(["No source position given"]))
        )
}

File: python-fp/app/chess/ChessBoard.py
from dataclasses import dataclass
from enum import Enum
from typing import List, Tuple, Callable, Iterable

from toolz import compose, curry

from app.chess.Color import Color, other
from app.chess.Position import charRangeInclusive, Position, toIndex, valid
from app.chess.utils import mapl, concat, chunks, Maybe, Nothing, Just, filterl, updatedList, Either, Left, Right, \
    Infix, lazy


class PieceType(Enum):
    Pawn = 1
    Knight = 2
    Bishop = 3
    Rook = 4
    Queen = 5
    King = 6


@dataclass
class Piece:
    color: Color
    type: PieceType

    def __iter__(self):
        return [self.color, self.type].__iter__()

    def __str__(self):
        if self.color == Color.WHITE:
            return {PieceType.Pawn: "♙",
                    PieceType.Knight: "♘",
                    PieceType.Bishop: "♗",
                    PieceType.Rook: "♖",
                    PieceType.Queen: "♕",
                    PieceType.King: "♔"
                    }[self.type]
        else:
            return {PieceType.Pawn: "♟",
                    PieceType.Knight: "♞",
                    PieceType.Bishop: "♝",
                    PieceType.Rook: "♜",
                    PieceType.Queen: "♛",
                    PieceType.King: "♚"
                    }[self.type]


ANSI_RESET = "\u001B[0m"


class BackgroundColor(Enum):
    Yellow = 1
    Maroon = 2
    Cyan = 3
    White = 4


def ansiCode(c: BackgroundColor) -> str:
    return {
        BackgroundColor.Yellow: "\u001B[43m",
        BackgroundColor.Maroon: "\u001B[41m",
        BackgroundColor.Cyan: "\u001B[47m",
        BackgroundColor.White: "\u001B[46m"
    }[c]


class ChessBoard:
    """A structure that can have any fields defined."""

    def __init__(self, **entries): self.__dict__.update(entries)

    def toVector(self) -> List[Maybe[Piece]]:
        raise Exception("unimplemented")

    def highlights(self) -> List[Maybe[BackgroundColor]]:
        raise Exception("unimplemented")

    def nextPlayer(self) -> Color:
        raise Exception("unimplemented")


def switch(cb: ChessBoard) -> ChessBoard:
    return ChessBoard(toVector=lambda: cb.toVector(),
                      highlights=lambda: cb.highlights(),
                      nextPlayer=lambda: other(cb.nextPlayer()))


def initialPosition() -> ChessBoard:
    rearRow = [PieceType.Rook, PieceType.Knight, PieceType.Bishop, PieceType.Queen, PieceType.King, PieceType.Bishop,
               PieceType.Knight, PieceType.Rook]
    whiteRearRow = list(map(lambda typ: Just(Piece(Color.WHITE, typ)), rearRow))
    whiteFrontRow = [Just(Piece(Color.WHITE, PieceType.Pawn)) for i in range(0, 8)]
    emptyRow = [Nothing for i in range(0, 8)]

    blackFrontRow = [Just(Piece(Color.BLACK, PieceType.Pawn)) for i in range(0, 8)]
    blackRearRow = list(map(lambda typ: Just(Piece(Color.BLACK, typ)), rearRow))

    cells = whiteRearRow + whiteFrontRow + \
            concat([emptyRow for i in range(0, 4)]) + \
            blackFrontRow + blackRearRow

    return ChessBoard(toVector=lambda: cells,
                      highlights=lambda: [Nothing for i in range(0, 64)],
                      nextPlayer=lambda: Color.WHITE)


def showChessBoard(cb: ChessBoard) -> str:
    columnHeader = "".join(mapl(lambda x: f" {x} ", charRangeInclusive('a', 'h')))
    rowHeader = [f" {i + 1} " for i in range(0, 8)]
    alternateBlackWhite = concat([concat([[BackgroundColor.White, BackgroundColor.Cyan] for i in range(0, 4)]) +
                                  concat([[BackgroundColor.Cyan, BackgroundColor.White] for i in range(0, 4)])
                                  for i in range(0, 4)
                                  ])

    colorCodes = list(zip(cb.highlights(), alternateBlackWhite))
    cells = list(zip(colorCodes, cb.toVector()))

    cells2d = list(chunks(cells, 8))

    def showLine(highlightCol_bgCol_piece: Tuple[Tuple[Maybe[BackgroundColor], BackgroundColor], Maybe[Piece]]):
        (colors, piece) = highlightCol_bgCol_piece
        (highlightCol, bgCol) = colors
        return ansiCode(highlightCol.orElse(bgCol)) + " " + str(piece.orElse(" ")) + " " + ANSI_RESET

    lines = mapl(lambda lineCells: "".join(mapl(showLine, lineCells)), cells2d)

    def concatTuples(xy: Tuple[str, str]):
        return str(xy[0]) + str(xy[1]) + str(xy[0])

    return f"   {columnHeader}\n" + "\n".join(mapl(concatTuples, zip(rowHeader, lines))) + f"\n   {columnHeader}"


# print(showChessBoard(initialPosition()))


def at(cb: ChessBoard, pos: Position) -> Maybe[Piece]:
    i = toIndex(pos)
    if i < 0 or i > 63:
        return Nothing
    return cb.toVector()[i]


def cellIsEmpty(cb: ChessBoard, pos: Position):
    return at(cb, pos) == Nothing


def diffColor(cb: ChessBoard, x: Position, y: Position) -> bool:
    def match(px: Piece, py: Piece):
        (colX, typX) = px
        (colY, typY) = py
        return colX != colY

    return at(cb, x).map(lambda posX:
                         at(cb, y).map(lambda posY: match(posX, posY)).orElse(False)
                         ).orElse(False)


def movesRecursion(cb: ChessBoard, start: Position, here: Position, fn: Callable) -> List[Position]:
    nextt = fn(here)
    if not valid(nextt):
        return []
    if cellIsEmpty(cb, nextt):
        return [nextt] + movesRecursion(cb, start, nextt, fn)
    if diffColor(cb, start, nextt):
        return [nextt]
    return []


def movesFrom(cb: ChessBoard, here: Position, fn: Callable) -> List[Position]:
    return movesRecursion(cb, here, here, fn)


def up(pos: Position) -> Position:
    (i, j) = pos
    return i - 1, j


def down(pos: Position) -> Position:
    (i, j) = pos
    return i + 1, j


def left(pos: Position) -> Position:
    (i, j) = pos
    return i, j - 1


def right(pos: Position) -> Position:
    (i, j) = pos
    return i, j + 1


def diagUpLeft(pos: Position) -> Position:
    (i, j) = pos
    return i - 1, j - 1


def diagUpRight(pos: Position) -> Position:
    (i, j) = pos
    return i - 1, j + 1


def diagDownLeft(pos: Position) -> Position:
    (i, j) = pos
    return i + 1, j - 1


def diagDownRight(pos: Position) -> Position:
    (i, j) = pos
    return i + 1, j + 1


def rookMoves(cb: ChessBoard, pos: Position) -> List[Position]:
    return movesFrom(cb, pos, up) + \
           movesFrom(cb, pos, down) + \
           movesFrom(cb, pos, left) + \
           movesFrom(cb, pos, right)


def take(n: int, items: Iterable):
    return [x for _, x in zip(range(n), items)]


def kingMoves(cb: ChessBoard, pos: Position) -> List[Position]:
    return take(1, movesFrom(cb, pos, up)) + \
           take(1, movesFrom(cb, pos, down)) + \
           take(1, movesFrom(cb, pos, left)) + \
           take(1, movesFrom(cb, pos, right)) + \
           take(1, movesFrom(cb, pos, diagUpLeft)) + \
           take(1, movesFrom(cb, pos, diagUpRight)) + \
           take(1, movesFrom(cb, pos, diagDownLeft)) + \
           take(1, movesFrom(cb, pos, diagDownRight))


def bishopMoves(cb: ChessBoard, pos: Position) -> List[Position]:
    return movesFrom(cb, pos, diagUpLeft) + \
           movesFrom(cb, pos, diagUpRight) + \
           movesFrom(cb, pos, diagDownLeft) + \
           movesFrom(cb, pos, diagDownRight)


def queenMoves(cb: ChessBoard, pos: Position) -> List[Position]:
    return rookMoves(cb, pos) + bishopMoves(cb, pos)


AND = Infix(lambda f, g: (lambda x: f(x) and g(x)))


def knightMoves(cb: ChessBoard, pos: Position) -> List[Position]:
    fn = lambda y: diffColor(cb, pos, y) or cellIsEmpty(cb, y)

    return filterl(valid |AND| fn,
                   [compose(up, up, left)(pos)]
                   + [compose(up, up, right)(pos)]
                   + [compose(down, down, left)(pos)]
                   + [compose(down, down, right)(pos)]
                   + [compose(left, left, up)(pos)]
                   + [compose(left, left, down)(pos)]
                   + [compose(right, right, up)(pos)]
                   + [compose(right, right, down)(pos)]
                   )


def pawnMovesNonCapture(cb: ChessBoard, pos: Position, dirFn: Callable, movingForTheFirstTime: Callable) -> List[Position]:
    isEmpty = lambda p: cellIsEmpty(cb, p)

    nextt = dirFn(pos)
    nextToNext = compose(dirFn, dirFn)(pos)

    if movingForTheFirstTime(pos):
        return [nextt for _ in range(0, 1) if isEmpty(nextt)] + \
               [nextToNext for _ in range(0, 1) if isEmpty(nextt) and isEmpty(nextToNext)]

    return [nextt for _ in range(0, 1) if isEmpty(nextt)]


def pawnMovesDown(cb: ChessBoard, pos: Position) -> List[Position]:
    return pawnMovesNonCapture(cb, pos, down, lambda p: p[0] == 1) + \
           filterl(lambda y: diffColor(cb, pos, y), [diagDownLeft(pos)]) + \
           filterl(lambda y: diffColor(cb, pos, y), [diagDownRight(pos)])


def pawnMovesUp(cb: ChessBoard, pos: Position) -> List[Position]:
    return pawnMovesNonCapture(cb, pos, up, lambda p: p[0] == 6) + \
           filterl(lambda y: diffColor(cb, pos, y), [diagUpLeft(pos)]) + \
           filterl(lambda y: diffColor(cb, pos, y), [diagUpRight(pos)])


def possiblePositionsToMove(cb: ChessBoard, pos: Position) -> List[Position]:
    def getMoves(piece: Piece):
        (col, typ) = piece
        return {
            PieceType.Rook: lambda: rookMoves(cb, pos),
            PieceType.Knight: lambda: knightMoves(cb, pos),
            PieceType.Bishop: lambda: bishopMoves(cb, pos),
            PieceType.King: lambda: kingMoves(cb, pos),
            PieceType.Queen: lambda: queenMoves(cb, pos),
            PieceType.Pawn: lambda: pawnMovesDown(cb, pos) if col == Color.WHITE else pawnMovesUp(cb, pos)
        }.get(typ, lazy([]))()

    return at(cb, pos).map(getMoves) \
        .orElse([])


def showPossibleMoves(pos: Maybe[Position], cb: ChessBoard) -> ChessBoard:
    def highlightedCells(p: Position):
        cells = mapl(lambda x: toIndex(x), possiblePositionsToMove(cb, p))
        return [Just(BackgroundColor.Yellow) if x in cells else Nothing for x in range(0, 64)]

    return pos.map(lambda p: ChessBoard(toVector=lambda: cb.toVector(),
                                        highlights=lambda: highlightedCells(p),
                                        nextPlayer=lambda: cb.nextPlayer())) \
        .orElse(cb)


def movePiece(posX: Position, posY: Position, cb: ChessBoard) -> ChessBoard:
    def updatedCells(x: int, y: int, pieces: List[Maybe[Piece]]):
        return updatedList(pieces, [(y, pieces[x]), (x, Nothing)])

    return ChessBoard(toVector=lambda: updatedCells(toIndex(posX), toIndex(posY), cb.toVector()),
                      highlights=lambda: [Nothing for _ in range(0, 64)],
                      nextPlayer=lambda: other(cb.nextPlayer()))


def validateAndMakeMove(posX: Maybe[Position], posY: Maybe[Position], cb: ChessBoard) -> Either[List[str], ChessBoard]:
    @curry
    def ifValidTargetFinallyMoveFrom(px: Position, py: Position):
        ys = mapl(lambda p: toIndex(p), possiblePositionsToMove(cb, px))
        return Right(movePiece(px, py, cb)) if toIndex(py) in ys else Left(["This move not allowed!"])

    @curry
    def ifMyTurn(px: Position, piece: Piece):
        (col, typ) = piece
        if not (col == cb.nextPlayer()):
            return Left(["Not your turn!"])

        return posY\
                .map(ifValidTargetFinallyMoveFrom(px))\
                .orElse(Left([f"Target position not given"]))

    def maybeMoveFromSource(x: Position):
        return at(cb, x).map(ifMyTurn(x)).orElse(Left(["No piece at given position"]))

    return posX.map(maybeMoveFromSource) \
        .orElse(Left(["No source position given!"]))
